Adds live migration procedures for sdn to ovnk (#76672)
openshift/openshift-docs: modules/nw-ovn-kubernetes-live-migration.adoc

// Module included in the following assemblies:
//
// * networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc
// * networking/openshift_sdn/rollback-to-ovn-kubernetes.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ovn-kubernetes-live-migration_{context}"]
= Migrating to the OVN-Kubernetes network plugin by using the live migration method

The following procedure entails checking for egress router resources and using the live migration method to migrate from the OpenShift SDN network plugin to the OVN-Kubernetes network plugin.

.Prerequisites

* A cluster has been configured with the OpenShift SDN CNI network plugin in the network policy isolation mode.
* You have installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.

> * You have created a recent backup of the etcd database is available.

Reviewer: ```suggestion
* You have created a recent backup of the etcd database
```
or
```suggestion
* You have a recent backup of the etcd database available.
```

* The cluster is in a known good state without any errors.

> * On all cloud platforms after updating software, a security group rule must be in place to allow UDP packets on port `6081` for all nodes.

Reviewer: My ignorance: what software?

> On all cloud platforms after updating software,

Reviewer 2: I suppose it refers to OpenShift.

+
[source,terminal]
----
{
  "name": "egress-multi",
  "namespace": "egress-router-project"
}
----
+
** Alternatively, you can query metrics on the {product-title} web console or by using the `oc` CLI to check for egress router pods. For more information, see "Checking live migration metrics".

. Enter the following command to remove an egress router pod:
+
[source,terminal]
----
$ oc delete pod <egress_pod_name> -n <egress_router_project>
----

ifdef::openshift-rosa,openshift-dedicated[]
. Enter the following command to add the `unsupported-red-hat-internal-testing` annotation to the cluster-level network configuration:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"unsupported-red-hat-internal-testing": "true"}}}'
----
endif::[]

. Enter the following command to patch the cluster-level networking configuration and initiate the migration from OpenShift SDN to OVN-Kubernetes:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OVNKubernetes"}}'
----
+

> After running these commands, the migration process begins. During this process, two rounds of node reboot are triggered with the Machine Config Operator. The expected time that this process takes is approximately two times that of a cluster upgrade.

Reviewer: ```suggestion
After running these commands, the migration process begins. During this process, the Machine Config Operator reboots the nodes in your cluster twice. It is expected that the migration takes approximately twice as long as a cluster upgrade.
```

+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"unsupported-red-hat-internal-testing": "true"}}}'
----
endif::[]

. Enter the following command to patch the cluster-level networking configuration and initiate the migration from OpenShift SDN to OVN-Kubernetes:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OVNKubernetes"}}'
----
+
After running these commands, the migration process begins. During this process, two rounds of node reboot are triggered with the Machine Config Operator. The expected time that this process takes is approximately two times that of a cluster upgrade.

. Optional: You can enter the following commands to ensure that the migration process has completed, and to check the status of the `network.config`:
+
[source,terminal]
----
$ oc get network.config.openshift.io cluster -o jsonpath='{.status.networkType}'
----
+
[source,terminal]
----
$ oc get network.config cluster -o=jsonpath='{.status.conditions}' | jq .
----
+

> You can check live migration metrics for troubleshooting issues. For more information, see "Checking live migration metrics".

Reviewer 2: Once, the live migration is completed. Users shall remove the annotation `network.openshift.io/network-type-migration` from the network.config CR.

Author: @pliurh added the following:

. Complete the following steps only if the migration succeeds and your cluster is in a good state:

.. To remove the migration configuration from the `network.config` custom resource, enter the following command:
+
[source,terminal]
----
$ oc patch Network.operator.openshift.io cluster --type='merge' \
  --patch '{ "spec": { "migration": null } }'
----

.. To remove custom configuration for the OpenShift SDN network provider, enter the following command:
+
[source,terminal]
----
$ oc patch Network.operator.openshift.io cluster --type='merge' \
  --patch '{ "spec": { "defaultNetwork": { "openshiftSDNConfig": null } } }'
----

.. To remove the OpenShift SDN network provider namespace, enter the following command:
+
[source,terminal]
----
$ oc delete namespace openshift-sdn
----

:_mod-docs-content-type: PROCEDURE
[id="nw-ovn-kubernetes-live-migration_{context}"]
= Migrating to the OVN-Kubernetes network plugin by using the live migration method

The following procedure entails checking for egress router resources and using the live migration method to migrate from the OpenShift SDN network plugin to the OVN-Kubernetes network plugin.

.Prerequisites

* A cluster has been configured with the OpenShift SDN CNI network plugin in the network policy isolation mode.
* You have installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.
* You have created a recent backup of the etcd database is available.
* The cluster is in a known good state without any errors.
* On all cloud platforms after updating software, a security group rule must be in place to allow UDP packets on port `6081` for all nodes.
* Cluster administrators must remove any egress router resources.
* You have reviewed the "Considerations for live migration to the OVN-Kubernetes network plugin" section of this document.

.Procedure 

. On the {product-title} web console, click *Observe* -> *Metrics*.

> . In the *Expression* box, add the following queries, then click *Add query*:

Reviewer 3: live migration has pre-validate if there is egressrouter running on cluster,  So this query is no need I think ,we can remove it. 

this one can be replaced with :

if there is egressrouter feature on the cluster, when do migration, the network operator will blocked the migration and expose one error "
The cluster configuration is invalid (network type live migration is not supported for pods with "pod.network.openshift.io/assign-macvlan" annotation. Please remove all egress router pods). Use 'oc edit network.config.openshift.io cluster' to fix"

And also we can see the metric query `openshift_network_operator_live_migration_blocked`, there is `UnsupportedMACVLANInterface` from console. 

or we can query from cli console. 
```
#token=`oc create token prometheus-k8s -n openshift-monitoring`
# oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query?" --data-urlencode "query=openshift_network_operator_live_migration_blocked" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   515  100   460  100    55   9583   1145 --:--:-- --:--:-- --:--:-- 10729
{
  "status": "success",
  "data": {
    "resultType": "vector",
    "result": [
      {
        "metric": {
          "__name__": "openshift_network_operator_live_migration_blocked",
          "container": "network-operator",
          "endpoint": "metrics",
          "instance": "10.0.3.18:9104",
          "job": "metrics",
          "namespace": "openshift-network-operator",
          "pod": "network-operator-5f9874cc47-676rg",
          "prometheus": "openshift-monitoring/k8s",
          "reason": "UnsupportedMACVLANInterface",
          "service": "metrics"
        },
        "value": [
          1717558701.173,
          "1"
        ]
      }
    ],
    "analysis": {}
  }
}
```

+
[source,text]
----
network_attachment_definition_instances{networks="egress-router"} > 0
----

. Click *Run queries*. If any data points are found, you should not proceed with the OpenShift SDN to OVN-Kubernetes live migration procedure. These custom resources must be removed before using the live migration procedure.

. To remove an egress router, you can delete the egress router custom resource definition:
+
[source,terminal]
----

> $ oc delete crd egressrouters.network.operator.openshift.io

Reviewer 3: do we need to remove resource egressrouters CRD?  I think remove the egress pod is enough

Reviewer 2: If we only remove the CRD, wouldn't the pod be created again before we start the migration?

Author: I think that this is resolved with Step 1 cc @pliurh @zhaozhanqi

----

ifdef::openshift-rosa,openshift-dedicated[]

> . Enter the following command to add the `unsupported-red-hat-internal-testing` annotation to the cluster-level network configuration:

Reviewer 2: Just for clarification, this is only needed by managed clusters.

Author: Got it. The ifdef statements will ensure that this step is only in ROSA/OSD clusters. 

@pliurh regarding `$ oc delete crd egressrouters.network.operator.openshift.io`, this is what we did in the KCS. Is there another or different procedure required?

Reviewer 2: The `egressrouters` is the OVN-K API for configuring egress router. In OpenShiftSDN, we use egress router pods instead. So we shall ask users to remove all the egress router pods before migration. https://docs.openshift.com/container-platform/4.14/networking/openshift_sdn/deploying-egress-router-layer3-redirection.html

Author: Can you check Step 1 and ensure its accuracy? cc @zhaozhanqi because it addresses https://github.com/openshift/openshift-docs/pull/76672#discussion_r1630962831 as well.

Reviewer 3: ok for me
and maybe we can also provide how to find the egressrouter pod `oc get pods --all-namespaces -o json | jq '.items[] | select(.metadata.annotations."pod.network.openshift.io/assign-macvlan" == "true") | {name: .metadata.name, namespace: .metadata.namespace}'`

Author: @zhaozhanqi Updated to add that command here: https://76672--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.html#nw-ovn-kubernetes-live-migration_migrate-from-openshift-sdn

Should we tell users how to delete those pods here?

+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"unsupported-red-hat-internal-testing": "true"}}}'
----
endif::[]

. Enter the following command to patch the cluster-level networking configuration and initiate the migration from OpenShift SDN to OVN-Kubernetes:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OVNKubernetes"}}'
----
+

> After running these commands, the migration process begins. The time it takes to complete the process will vary depending on how many nodes are in your cluster. 

Reviewer 2: Add
> During the migration process, we will trigger 2 rounds of node reboot with the Machine Config Operator. So the expected time it takes is about 2 times of a cluster upgrade.

Author: Added

> . Optional: You can enter the following commands to ensure that the migration process has completed, and to check the status of the `network.config`:

Reviewer 2: We added new metrics for users to monitor the migration progress in 4.16. https://docs.google.com/document/d/1QQJDtipDqSyLnOYkoEJ12QnV-g61d6BwDLY-aOpdZVw/edit

Reviewer 3: above link show check the metrics from console. 

and we can also provide another way to check from oc client:

```
# token=`oc create token prometheus-k8s -n openshift-monitoring`
# oc -n openshift-monitoring exec -c prometheus prometheus-k8s-0 -- curl -k -H "Authorization: Bearer $token" "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query?" --data-urlencode "query=openshift_network_operator_live_migration_condition" | jq`
{
  "status": "success",
  "data": {
    "resultType": "vector",
    "result": [
      {
        "metric": {
          "__name__": "openshift_network_operator_live_migration_condition",
          "container": "network-operator",
          "endpoint": "metrics",
          "instance": "10.0.83.62:9104",
          "job": "metrics",
          "namespace": "openshift-network-operator",
          "pod": "network-operator-6c87754bc6-c8qld",
          "prometheus": "openshift-monitoring/k8s",
          "service": "metrics",
          "type": "NetworkTypeMigrationInProgress"
        },
        "value": [
          1717653579.587,
          "1"
        ]
      },
      {
        "metric": {
          "__name__": "openshift_network_operator_live_migration_condition",
          "container": "network-operator",
          "endpoint": "metrics",
          "instance": "10.0.83.62:9104",
          "job": "metrics",
          "namespace": "openshift-network-operator",
          "pod": "network-operator-6c87754bc6-c8qld",
          "prometheus": "openshift-monitoring/k8s",
          "service": "metrics",
          "type": "NetworkTypeMigrationMTUReady"
        },
        "value": [
          1717653579.587,
          "0"
        ]
      },
      {
        "metric": {
          "__name__": "openshift_network_operator_live_migration_condition",
          "container": "network-operator",
          "endpoint": "metrics",
          "instance": "10.0.83.62:9104",
          "job": "metrics",
          "namespace": "openshift-network-operator",
          "pod": "network-operator-6c87754bc6-c8qld",
          "prometheus": "openshift-monitoring/k8s",
          "service": "metrics",
          "type": "NetworkTypeMigrationOriginalCNIPurged"
        },
        "value": [
          1717653579.587,
          "0"
        ]
      },
      {
        "metric": {
          "__name__": "openshift_network_operator_live_migration_condition",
          "container": "network-operator",
          "endpoint": "metrics",
          "instance": "10.0.83.62:9104",
          "job": "metrics",
          "namespace": "openshift-network-operator",
          "pod": "network-operator-6c87754bc6-c8qld",
          "prometheus": "openshift-monitoring/k8s",
          "service": "metrics",
          "type": "NetworkTypeMigrationTargetCNIAvailable"
        },
        "value": [
          1717653579.587,
          "0"
        ]
      },
      {
        "metric": {
          "__name__": "openshift_network_operator_live_migration_condition",
          "container": "network-operator",
          "endpoint": "metrics",
          "instance": "10.0.83.62:9104",
          "job": "metrics",
          "namespace": "openshift-network-operator",
          "pod": "network-operator-6c87754bc6-c8qld",
          "prometheus": "openshift-monitoring/k8s",
          "service": "metrics",
          "type": "NetworkTypeMigrationTargetCNIInUse"
        },
        "value": [
          1717653579.587,
          "0"
        ]
      }
    ],
    "analysis": {}
  }
}
```
